dataplane/linux/bpf_ep_mgr.go

```diff
@@ -496,44 +496,48 @@ func (m *bpfEndpointManager) applyPolicy(wlID proto.WorkloadEndpointID) error {
 	return nil
 }
 
-func (m *bpfEndpointManager) ensureQdisc(ifaceName string) {
-	tc.EnsureQdisc(ifaceName)
-}
-
 var calicoRouterIP = net.IPv4(169, 254, 1, 1).To4()
 
 func (m *bpfEndpointManager) attachWorkloadProgram(endpoint *proto.WorkloadEndpoint, polDirection PolDirection) error {
 	ap := m.calculateTCAttachPoint(tc.EpTypeWorkload, polDirection, endpoint.Name)
 	// Host side of the veth is always configured as 169.254.1.1.
-	ap.IP = calicoRouterIP
+	ap.HostIP = calicoRouterIP
 	// * VXLAN MTU should be the host ifaces MTU -50, in order to allow space for VXLAN.
 	// * We also expect that to be the MTU used on veths.
 	// * We do encap on the veths, and there's a bogus kernel MTU check in the BPF helper
 	//   for resizing the packet, so we have to reduce the apparent MTU by another 50 bytes
 	//   when we cannot encap the packet - non-GSO & too close to veth MTU
 	ap.TunnelMTU = uint16(m.vxlanMTU - 50)
-	err := ap.AttachProgram()
-	if err != nil {
-		return err
-	}
 
 	var tier *proto.TierInfo
 	if len(endpoint.Tiers) != 0 {
 		tier = endpoint.Tiers[0]
 	}
 	rules := m.extractRules(tier, endpoint.ProfileIds, polDirection)
 
-	jumpMapFD, err := FindJumpMap(ap)
-	if err != nil {
-		return errors.Wrap(err, "failed to look up jump map")
-	}
-	defer func() {
-		err := jumpMapFD.Close()
+	iface := m.ifaces[endpoint.Name]
+	if iface.jumpMapFD[polDirection] == 0 {
+		// We don't have a program attached to this interface yet, attach one now.
+		err := ap.AttachProgram()
 		if err != nil {
-			log.WithError(err).Panic("Failed to close jump map FD")
+			return err
 		}
-	}()
 
+		jumpMapFD, err := FindJumpMap(ap)
+		if err != nil {
+			return errors.Wrap(err, "failed to look up jump map")
+		}
+		if iface.jumpMapFD == nil {
+			iface.jumpMapFD = map[PolDirection]bpf.MapFD{}
+		}
+		iface.jumpMapFD[polDirection] = jumpMapFD
+		m.ifaces[endpoint.Name] = iface
+	}
+
+	return m.updatePolicyProgram(iface.jumpMapFD[polDirection], rules)
+}
+
+func (m *bpfEndpointManager) updatePolicyProgram(jumpMapFD bpf.MapFD, rules [][][]*proto.Rule) error {
```

Comment:
Is there any change needed here because of the jump map fd being held open?  E.g. if it was like a normal file, it would point to a particular offset in the file, and that might need resetting?

Reply:
No, I don't think BPF map FDs are stateful.